From b9da75196891496a7ee6e622711612389754b2ed Mon Sep 17 00:00:00 2001
From: Jamie Iles <jamie@jamieiles.com>
Date: Thu, 7 Apr 2011 12:09:55 +0100
Subject: [PATCH 073/228] picoxcell: provide a mechanism to register the TRNG

The TRNG (true random number generator) is available on pc3x3 and pc30xx
so provide a mechanism to register it dynamically.

Signed-off-by: Jamie Iles <jamie@jamieiles.com>
---
 arch/arm/mach-picoxcell/Makefile         |    3 ++-
 arch/arm/mach-picoxcell/device_trng.c    |   28 ++++++++++++++++++++++++++++
 arch/arm/mach-picoxcell/picoxcell_core.h |    1 +
 3 files changed, 31 insertions(+), 1 deletions(-)
 create mode 100644 arch/arm/mach-picoxcell/device_trng.c

diff --git a/arch/arm/mach-picoxcell/Makefile b/arch/arm/mach-picoxcell/Makefile
index 7fce320..561a0b7 100644
--- a/arch/arm/mach-picoxcell/Makefile
+++ b/arch/arm/mach-picoxcell/Makefile
@@ -8,7 +8,8 @@ obj-y				:= picoxcell_core.o io.o axi2cfg.o \
 				   bus_err.o \
 				   gpioirq.o \
 				   device_uart.o \
-				   device_spacc.o
+				   device_spacc.o \
+				   device_trng.o
 obj-$(CONFIG_BOARD_PC7302)	+= board_pc7302.o
 obj-$(CONFIG_PICOXCELL_PC3X2)	+= pc3x2.o
 obj-$(CONFIG_PICOXCELL_PC3X3)	+= pc3x3.o
diff --git a/arch/arm/mach-picoxcell/device_trng.c b/arch/arm/mach-picoxcell/device_trng.c
new file mode 100644
index 0000000..87f3da7
--- /dev/null
+++ b/arch/arm/mach-picoxcell/device_trng.c
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2011 Picochip Ltd., Jamie Iles
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * All enquiries to support@picochip.com
+ */
+#include <linux/err.h>
+#include <linux/platform_device.h>
+
+#include <mach/hardware.h>
+
+#include "picoxcell_core.h"
+
+int __init picoxcell_add_trng(unsigned long addr)
+{
+	struct resource res = {
+		.start		= addr,
+		.end		= addr + 0xFFFF,
+		.flags		= IORESOURCE_MEM,
+	};
+	struct platform_device *pdev =
+		platform_device_register_simple("picoxcell-trng", -1, &res, 1);
+
+	return pdev ? 0 : PTR_ERR(pdev);
+}
diff --git a/arch/arm/mach-picoxcell/picoxcell_core.h b/arch/arm/mach-picoxcell/picoxcell_core.h
index 960edbb..a762a1d 100644
--- a/arch/arm/mach-picoxcell/picoxcell_core.h
+++ b/arch/arm/mach-picoxcell/picoxcell_core.h
@@ -34,5 +34,6 @@ extern void __init armgpio_irq_init(void);
 extern int picoxcell_add_uart(unsigned long addr, int irq, int id);
 int __init picoxcell_add_spacc(const char *name, unsigned long addr, int irq,
 			       int id);
+int __init picoxcell_add_trng(unsigned long addr);
 
 #endif /* __ASM_ARCH_PICOXCELL_CORE_H__ */
-- 
1.7.0.4

